### PR TITLE
Add `output_symbols` arg to codegen as performance mitigation

### PR DESCRIPTION
**Pull request by gareth-cross**
For complex expressions, the calculation of `Codegen.output_symbols` can grow quite large (encroaching on a minute or more). In some cases, the value of `output_symbols` can be specified externally. For example:
- The user may have a list of `output_symbols` readily available at the callsite (frequently it matches the input symbol set exactly).
- The `output_symbols` can be determined from another `Codegen` object, for example in _most_ (though not all) cases the following can work:
```python
cg = Codegen(...)
cg_factor = cg.with_linearization(output_symbols=cg.output_symbols) # avoid recomputation
```

Ideally `free_symbols` would be faster, but this is a mitigation/workaround that can offer worthwhile savings in some cases.

This change adds an optional argument to the various constructors of `Codegen` so that the user can optionally provide the symbols themselves, and avoid an expensive computation.

### CODE REVIEW

File: symforce/codegen/codegen.py [hunk 1]
```diff
@@ -238,15 +249,10 @@ def __init__(
...
-    def output_symbols(self) -> T.Set[sf.Symbol]:
-        """
-        The set of free symbols in the output
-
-        Cached, because this is somewhat expensive to compute for large outputs
-        """
-        # Convert to Matrix before calling free_symbols because it's much faster to call once
-        return sf.S(sf.Matrix(codegen_util.flat_symbols_from_values(self.outputs)).mat).free_symbols
+    @staticmethod
+    def _compute_output_symbols(output_values: Values) -> T.Set[sf.Symbol]:
```
Comment (chao-qu-skydio) [inline symforce/codegen/codegen.py:253]: Prefer to add back the comments stating that this is slow.

File: symforce/codegen/codegen.py [hunk 2]
```diff
@@ -159,6 +164,12 @@ def __init__(
         self.inputs = inputs
         self.outputs = outputs
 
+        if output_symbols is None:
+            # compute the output symbols from the expressions:
```
Comment (chao-qu-skydio) [inline symforce/codegen/codegen.py:168]: Why do you end sentences in ":"? :crab:

**Review by chao-qu-skydio — COMMENTED** [review]
lgtm

File: symforce/codegen/codegen.py [hunk 3]
```diff
@@ -828,6 +838,7 @@ def with_jacobians(
         include_results: bool = True,
         name: str = None,
         sparse_jacobians: bool = False,
+        output_symbols: T.Optional[T.Set[sf.Symbol]] = None,
```
Comment (aaron-skydio) [inline symforce/codegen/codegen.py:841]: I think this one still needs to be forwarded to the `Codegen` constructor below?

**Pull request closed without merging**